### PR TITLE
Add feature to allow ceryx to be used as data by-pass proxy

### PR DESCRIPTION
**Pull request by w4tsn**
Hello there! I'm happy user of ceryx for around a year now and I've implemented a small set of features for a specific use-case at my startup: using ceryx as configurable proxy to by-pass a cluster application for database access.

The following merge request is rather big, but all the changes are important to enable this use-case / feature / pattern. Let me explain to you what is happening. I hope you can appreciate the additions and I'm happy to make adjustments wherever necessary.

The base material explaining the use-case on a rather theoretical basis is a document from the Eclipse Ditto documentation, [Data By-Pass Pattern](https://www.eclipse.org/ditto/advanced-data-by-pass.html). The basic idea is to place a random UUID with a "hidden" source URL together with authorization information in ceryx and return a 303 to the client which then is able to get a database query executed by calling ceryx with the UUID, without every seeing the actual query or credentials. The environment is a cluster application which is not suited to pipe big amounts of data like in a database query.

To enable this I've added several things which are all explained in more detail in the README.

- Ability to use arbitrary strings as source
- Ability to use request uri params for targets
- Ability to provide headers for ceryx calling the target, e.g. for Authorization or Host headers
- Ability to provide a TTL to automatically clean up and invalidate routes after a certain amount of time

As said previously I'm using this modified Ceryx for about a year now and I think these features could benefit more people out there.

Tell me what you think! :)

More ideas:

- Ceryx could execute queries in advance, cache them and discard after TTL runs out

### CODE REVIEW

**Comment by mgaerber** [comment]
👍

**Comment by parisk** [comment]
Hi @w4tsn thanks a lot for your proposal and your kind works. We will get back on this next week, in order to review the use case, semantics and implmentation.

**Review by parisk — CHANGES_REQUESTED** [review]
Hey! 👋 

I managed to find an hour during the weekend to review this Pull Request and think about the appropriate direction it should follow for the project. I also pushed a couple of commits, simplifying the documentation (https://github.com/sourcelair/ceryx/pull/89/commits/63ac665106f9050352552ba5ade2ec126030f5ec) and formatting code (https://github.com/sourcelair/ceryx/pull/89/commits/75ed15ae73df8d4f4f62434ce9803683c37da3b7).

First, we should abstract the behaviour around target headers and URL parameters. We have a great opportunity here to pave a straightforward path towards general header and URL parameter manipulation, in addition to just adding values.

Next, I was thinking of ways to keep code and flows as simple as possible — treat special cases as normal ones. In that spirit we should, we can avoid [JSON encoding and decoding](https://github.com/sourcelair/ceryx/pull/89/files#diff-116aaba3e746cdd1f6fa3fcbb2fb8523c21d53a5334543a944f35f940e221096R55-R56) by using an additional top-level key and reducing [multiple routing attempts](https://github.com/sourcelair/ceryx/pull/89/files#diff-763c70b3d07c188dcb36bf29662f96def53e8332fd7cadf62d86ce265fc4f619R68-R76) to one, by leveraging my first point.

## To do

To sum things up, we need a few things to get this merged

- [ ] Move `headers` from `settings` to top-level key.
    ```lua
    redis.prefix .. ":headers:" .. source
    ```
- [ ] Replace string values in `headers`, with an object:
    ```json
    {"name": "header-name", "action": "set", "value": "desired-header-value"}
    ```
- [ ] Move URL parameters to top-level key.
    ```lua
    redis.prefix .. ":parameters:" .. source
    ```
- [ ] Use objects for URL parameters, just like headers:
    ```json
    {"name": "parameter-name", "action": "set", "value": "desired-parameter-value"}
    ```

🙏 Once again, thanks a lot for your contribution @w4tsn. We really appreciate it and can't wait to get this merged upstream. 🚀

**Comment by w4tsn** [comment]
After thinking about this a bit more: I'm totally with you regarding your opinions and ideas on where to go from here but honestly I believe that all those requests is actually killing this MR because it extends the scope to something that I think is worth at least one more MR. Do you see any way of getting this MR merged and starting from there to tackle the refactor and new features?
